src/electos/ballotmaker/data/models.py

```diff
@@ -0,0 +1,220 @@
+
+    text: str
+    choices: List[BallotChoiceData]
+
+    def __post_init__(self):
+        super().__post_init__()
+        _check_type(self, "text", str)
+        _check_type_hint(self, "choices", List)
+        self.choices = [BallotChoiceData(**_) for _ in self.choices]
+
+
+# --- Candidate contest data models
+
+
+@dataclass
+class PartyData:
+
+    """Data for parties candidates are in."""
+
+    name: str
+    abbreviation: str
+
+    def __post_init__(self):
+        _check_type(self, "name", str)
+        _check_type(self, "abbreviation", str)
+
+
+@dataclass
+class CandidateChoiceData:
+
+    """Data for candidate contest selections."""
+
+    id: str
+    name: List[str]
+    party: List[PartyData]
+    is_write_in: bool
+
+    def __post_init__(self):
+        _check_type(self, "id", str)
+        _check_type_hint(self, "name", List)
+        _check_type_list(self, "name", str)
+        _check_type_hint(self, "party", List)
+        self.party = [PartyData(**_) for _ in self.party]
+        _check_type(self, "is_write_in", bool)
+
+
+@dataclass
+class CandidateContestData(ContestData):
+
+    """Data for candidate contests."""
+
+    vote_type: str
+    votes_allowed: str
+    candidates: List[CandidateChoiceData]
+
+    def __post_init__(self):
+        super().__post_init__()
+        _check_type(self, "vote_type", str)
+        _check_type(self, "votes_allowed", int)
+        _check_type_hint(self, "candidates", List)
+        self.candidates = [CandidateChoiceData(**_) for _ in self.candidates]
+
+
+@dataclass
+class BallotStyleData:
+
+    """Date for contests associated with a ballot style."""
+
+    # Note: Don't use separate fields for the types of contests.
+    # There's no guarantee the types will be clearly separated in an EDF.
+    # (The NIST SP-1500-100 JSON Schema uses unions too.)
+
+    id: str
+    scopes: List[str]
+    contests: List[Union[BallotMeasureContestData, CandidateContestData]]
+
+    def __post_init__(self):
+        _check_type(self, "id", str)
+        _check_type_hint(self, "scopes", List)
+        _check_type_list(self, "scopes", str)
+        _check_type_hint(self, "contests", List)
+        contests = []
+        contest_types = [_.value for _ in ContestType]
+        for contest in self.contests:
+            if not isinstance(contest, dict):
+                raise TypeError(f"Contest is not a dictionary: '{contest}'")
+            if "type" not in contest:
+                raise KeyError(f"Contest has no 'type' field: '{contest}'")
+            if contest["type"] not in contest_types:
+                raise ValueError(f"Unhandled contest type: '{contest['type']}'")
+            if contest["type"] == ContestType.BALLOT_MEASURE.value:
+                contest = BallotMeasureContestData(**contest)
+            elif contest["type"] == ContestType.CANDIDATE.value:
+                contest = CandidateContestData(**contest)
+            contests.append(contest)
```

Comment:
Does this mean that Ballot Measure Contests and Candidate Contests could be intermingled in the contests list? Does that matter?

Reply:
Not only can they be intermingled they _have_ to be. Splitting them up would lose the order they appear in the EDF.

However we can have our cake and eat it too. See the cached properties [`BallotStyleData.ballot_measure_contests`](https://github.com/ion-oset/ballot-lab/blob/edf-to-ballot-data/src/electos/ballotmaker/data/models.py#L185) and [`BallotStyleData.candidate_contests`](https://github.com/ion-oset/ballot-lab/blob/edf-to-ballot-data/src/electos/ballotmaker/data/models.py#L192) which extract them back out when you need them.